drivers/ipm/ipm_esp32.c

```diff
@@ -0,0 +1,280 @@
+			}
+			break;
+	}
+
+	/* unlock the shared memory */
+	atomic_set(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL);	
+}
+
+static int esp32_ipm_send(const struct device *dev, int wait, uint32_t id,
+			 const void *data, int size)
+{
+	size_t channel_size = esp32_ipm_channel_data_size(id);
+	struct esp32_ipm_data *dev_data = (struct esp32_ipm_data *)dev->data;
+
+	if(data == NULL) {
+		LOG_ERR("Invalid data source");
+		return -EINVAL;
+	}
+
+	if(!channel_size) {
+		LOG_ERR("Invalid IPM channel");
+		return -EINVAL;
+	}
+
+	if(channel_size < size) {
+		LOG_ERR("Not enough memory in IPM channel");
+		return -ENOMEM;
+	}
+
+	/* try to lock the shared memory */
+	while(!atomic_cas(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL, dev_data->this_core_id)) {
+		
+		k_busy_wait(1);
+
+		if((wait != -1) && (wait > 0)) {
+			/* lock could not be held this time, return */
+			wait--;
+			if(wait == 0)  {
+				return -ETIMEDOUT;	
+			}
+		}
+	}
+
+	/* We own the shared memory, so write data to it: */
+	esp32_ipm_copy_data_to_memory(dev_data->shm, id, data, size);
+	dev_data->control->dest_cpu_msg_id[dev_data->other_core_id] = id;
+
+	/* data copied, set the id and, generate interrupt in the remote core */
+	if (dev_data->this_core_id == 0) {
+		LOG_DBG("Generating interrupt on remote CPU 1 from CPU 0");
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_1_REG, DPORT_CPU_INTR_FROM_CPU_1);	
+	} else {
+		LOG_DBG("Generating interrupt on remote CPU 0 from CPU 1");
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, DPORT_CPU_INTR_FROM_CPU_0);
+	}
+
+	/* release the lock */
+	atomic_set(&dev_data->control->lock, ESP32_IPM_LOCK_FREE_VAL);
+
+	return 0;
+}
+
+static void esp32_ipm_register_callback(const struct device *dev,
+				       ipm_callback_t cb,
+				       void *user_data)
+{
+	struct esp32_ipm_data *data = (struct esp32_ipm_data *)dev->data;
+	uint32_t id; 
+
+	if(user_data == NULL) {
+		id = 0;
+	} else {
+		id = (uint32_t )user_data;
+	}
+
+	/* bind the id to the highest channel which is the user 1,
+	 * since this api does not return error
+	 */
+	if(id > DT_PROP(DT_NODELABEL(ipm0), ipm_channels) - 1){
+		LOG_WRN("Invalid ID assignment, defaults it to user channel");
+		id = DT_PROP(DT_NODELABEL(ipm0), ipm_channels) - 1;
+	} 
+
+	data->cb[id] = cb;
+}
+
+static int esp32_ipm_init(const struct device *dev)
+{
+	int irq_source;
+	struct esp32_ipm_data *data = (struct esp32_ipm_data *)dev->data;
+	
+	data->this_core_id = esp32_core_id();
+	data->other_core_id = (data->this_core_id  == 0) ? 1 : 0;
+	data->shm = (struct esp32_ipm_memory *)DT_REG_ADDR(DT_NODELABEL(shm0));
+	data->control = (struct esp32_ipm_control *)DT_REG_ADDR(DT_NODELABEL(ipm0));
+
+	LOG_DBG("Address of IPM shared memory: %p", data->shm);
+	LOG_DBG("Address of IPM control structure: %p", data->control);
+	
+	/* pro_cpu is responsible to initialize the lock of shared memory */
+	if(data->this_core_id == 0) {
+		irq_source = DT_IRQN(DT_NODELABEL(ipi0));
```

Comment:
same here. Can you reference `ipi0` and `ipi1` from DT instead of hardcoding?

Reply:
The ipi0 and ipi1 are hardwired to cpu number they belong, I don't see any advantage to remove the hardcode here.